src/ucp/core/ucp_worker.c

```diff
@@ -2118,10 +2111,19 @@ ucs_status_t ucp_worker_get_ep_config(ucp_worker_h worker,
         return UCS_ERR_EXCEEDS_LIMIT;
     }
 
-    ep_config = ucs_array_append_safe(ep_config_arr, &worker->ep_config,
-                                      ucp_worker_ep_config_array_free_cb, worker,
-                                      return UCS_ERR_NO_MEMORY);
-    status    = ucp_ep_config_init(worker, ep_config, key);
+    old_ep_cfg_buf = NULL;
+    ep_config      = ucs_array_append_safe(ep_config_arr, &worker->ep_config,
+                                           &old_ep_cfg_buf,
+                                           return UCS_ERR_NO_MEMORY);
+    if (old_ep_cfg_buf != NULL) {
+        /* Schedule release of old ep configs array backing buffer on the main
+         * thread (this func can be called by async thread).
```

Comment:
can you elaborate? Just add this to the end of comment? (does not look clear if yes)

Reply:
yes, add it to the end of the comment, to explain why returning a pointer to the old buffer makes it thread-safe
usually when thinking about thread-safe data struct it means it can be freely used from multiple threads at the same time, using some kind of sync/lock, and this is not the case here

Reply:
```
Schedule release of old ep configs array backing buffer on the main thread (this func can be called by async thread).
So the main thread can still access the old configuration buffer if it's modified by async thread.
```